src/Rule/Date/BaseDate.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator\Rule\Date;
+
+use Closure;
+use DateTimeInterface;
+use IntlDateFormatter;
+use Yiisoft\Validator\Rule\Trait\SkipOnEmptyTrait;
+use Yiisoft\Validator\Rule\Trait\SkipOnErrorTrait;
+use Yiisoft\Validator\Rule\Trait\WhenTrait;
+use Yiisoft\Validator\RuleInterface;
+use Yiisoft\Validator\SkipOnEmptyInterface;
+use Yiisoft\Validator\SkipOnErrorInterface;
+use Yiisoft\Validator\WhenInterface;
+
+/**
+ * @see BaseDateHandler
+ *
+ * @psalm-type IntlDateFormatterFormat = IntlDateFormatter::FULL | IntlDateFormatter::LONG | IntlDateFormatter::MEDIUM | IntlDateFormatter::SHORT | IntlDateFormatter::NONE
+ * @psalm-import-type SkipOnEmptyValue from SkipOnEmptyInterface
+ * @psalm-import-type WhenType from WhenInterface
+ */
+abstract class BaseDate implements RuleInterface, SkipOnErrorInterface, WhenInterface, SkipOnEmptyInterface
+{
+    use SkipOnEmptyTrait;
+    use SkipOnErrorTrait;
+    use WhenTrait;
+
+    /**
+     * @psalm-param non-empty-string|null $timeZone
+     * @psalm-param SkipOnEmptyValue $skipOnEmpty
+     * @psalm-param WhenType $when
+     */
+    public function __construct(
+        private ?string $format,
```

Comment:
Can `$format` include special symbols #, ?, * and etc?

https://www.php.net/manual/ru/datetimeimmutable.createfromformat.php

Reply:
If format started from "php:", it pass as format to `DateTimeImmutable::createFromFormat`, otherwise it pass as pattern to `IntlDateFormatter` constructor.

Reply:
In the case this will work incorrect
https://github.com/yiisoft/validator/pull/678/files#diff-db45265e469dc9c99923a3a89bd47256e93d65c5e1a34b8e108f22c4a07befe6R191

Reply:
Could you add test with date format `Y-*-d`?

Reply:
> In the case this will work incorrect https://github.com/yiisoft/validator/pull/678/files#diff-db45265e469dc9c99923a3a89bd47256e93d65c5e1a34b8e108f22c4a07befe6R191

It's other. Date rule has two formats:
- `format` for parsing value
- `messageFormat` for print value in error message

What you means? parsing or formatting?

Reply:
It seems this is for the message format and there is no problem with this.